bokehjs/src/lib/models/ranges/factor_range.ts

```diff
@@ -97,14 +119,159 @@ export function map_three_levels(
   let total_subpad = 0
   for (const [f0, top] of tops) {
     const n = top.length
-    const [submap, subpad] = map_two_levels(top, inner_pad, factor_pad, suboffset)
-    total_subpad += subpad
-    const subtot = sum(top.map(([f1]) => submap.get(f1)!.value))
-    mapping.set(f0, {value: subtot/n, mapping: submap})
-    suboffset += n + outer_pad + subpad
+    const sub = map_two_levels(top, inner_pad, factor_pad, suboffset)
+    total_subpad += sub.inner_padding
+    const subtot = sum(top.map(([f1]) => sub.mapping.get(f1)!.value))
+    mapping.set(f0, {value: subtot/n, mapping: sub.mapping})
+    suboffset += n + outer_pad + sub.inner_padding
+  }
+
+  const mids: L2Factor[] = []
+  for (const [f0, L2] of mapping) {
+    for (const f1 of L2.mapping.keys()) {
+      mids.push([f0, f1])
+    }
+  }
+
+  const inner_padding = (tops.size - 1)*outer_pad + total_subpad
+  return {mapping, tops: [...mapping.keys()], mids, inner_padding}
+}
+
+const is_l1 = (x: unknown) => isString(x)
+const is_l2 = (x: unknown) => isArray(x) && x.length == 2 && isString(x[0]) && isString(x[1])
+const is_l3 = (x: unknown) => isArray(x) && x.length == 3 && isString(x[0]) && isString(x[1]) && isString(x[2])
+
+export abstract class FactorMapper<FactorType> {
+  readonly levels: FactorLevel
+  readonly mids: L2Factor[] | null
+  readonly tops: L1Factor[] | null
+  readonly inner_padding: number
+
+  protected readonly mapping: MappingFor<FactorType>
+
+  constructor(
+    {levels, mapping, tops = null, mids = null, inner_padding}: {
+      levels: FactorLevel
+      mapping: MappingFor<FactorType>
+      tops?: L1Factor[] | null
+      mids?: L2Factor[] | null
+      inner_padding: number
+    },
+  ) {
+    this.levels = levels
+    this.mapping = mapping
+    this.tops = tops
+    this.mids = mids
+    this.inner_padding = inner_padding
+  }
+
+  static compute_levels(factors: Factor[]): FactorLevel {
+    if (every(factors, is_l1)) {
+      return 1
+    }
+    if (every(factors, is_l2)) {
+      return 2
+    }
+    if (every(factors, is_l3)) {
+      return 3
+    }
+    throw TypeError("factor levels are inconsistent")
+  }
+
+  static for(range: FactorRange): L1FactorMapper | L2FactorMapper | L3FactorMapper {
+    switch (this.compute_levels(range.factors)) {
+      case 1: {
+        return new L1FactorMapper(range)
+      }
+      case 2: {
+        return new L2FactorMapper(range)
+      }
+      case 3: {
+        return new L3FactorMapper(range)
+      }
+    }
+  }
+
+  map(x: FactorLike): number {
+    if (isNumber(x)) {
+      return x
+    }
+
+    const [boxed, offset] = (() => {
+      if (isString(x)) {
+        return [[x], 0]
+      }
+      const last = x.at(-1)
+      if (isNumber(last)) {
+        return [x.slice(0, -1), last]
+      }
+      return [x, 0]
+    })()
+
+    if (boxed.length > this.levels) {
+      throw TypeError(`Attempted to map ${boxed.length} levels of factors with an L${this.levels}FactorMap`)
```

Comment:
This looks like a value error rather a type error. Typically one would just throw `new Error("...")` anyway or use `assert()` if this is something that user shouldn't ever see (I suppose not the case here).